src/v/storage/index_state.cc

```diff
@@ -759,12 +759,21 @@ bool index_state::truncate(
     if (new_max_offset < base_offset) {
         return needs_persistence;
     }
-    const uint32_t i = new_max_offset() - base_offset();
-    auto res = index.offset_lower_bound(i);
-    size_t remove_back_elems = index.size() - res.value_or(index.size());
-    if (remove_back_elems > 0) {
-        needs_persistence = true;
-        pop_back(remove_back_elems);
+    static constexpr int64_t u32_max = std::numeric_limits<uint32_t>::max();
+    int64_t delta = new_max_offset() - base_offset();
+    // NOTE: we expect that deltas above u32_max would have not been
+    // added to the index, unless by an older buggy version of Redpanda.
+    // With this in mind, either there are no entries above u32_max, or
+    // offset_lower_bound() isn't going to work correctly anyway, so we just
+    // skip removal and rely on queries to detect the overflow.
```

Comment:
> so we just skip removal and rely on queries to detect the overflow.

i'm not clear on what this means or how it works. the confusion i have is something like: if the truncation request is legit, but we ignore it because of the overflow, then how do we know that a later query would not be affected by having dropped a truncation request? presumably we rely on suffix truncation for raft correctness?

Reply:
The idea is that if entries _aren't_ truncated here, it's because we are attempting to truncate to something beyond the uint32 delta limit. If that's the case:
- there [should be no such entries in the index](https://github.com/redpanda-data/redpanda/blob/32d02e68b56f1935f9bce9c1694ce0b3e2bf1d64/src/v/storage/index_state.cc#L305), and
- queries for beyond that limit [should skip the binary search](https://github.com/redpanda-data/redpanda/blob/32d02e68b56f1935f9bce9c1694ce0b3e2bf1d64/src/v/storage/index_state.cc#L688-L692)

Reply:
i think it looks ok. i'm less concerned about my question after read it again today